src/ripple/app/reporting/ETLSource.cpp

```diff
@@ -0,0 +1,919 @@
+{
+    org::xrpl::rpc::v1::GetLedgerResponse response;
+    bool success = execute(
+        [&response, ledgerSequence, getObjects, this](auto& source) {
+            auto [status, data] =
+                source->fetchLedger(ledgerSequence, getObjects);
+            response = std::move(data);
+            if (status.ok() and response.validated())
+            {
+                JLOG(journal_.info())
+                    << "Successfully fetched ledger = " << ledgerSequence
+                    << " from source = " << source->toString();
+                return true;
+            }
+            else
+            {
+                JLOG(journal_.warn())
+                    << "Error getting ledger = " << ledgerSequence
+                    << " Reply : " << response.DebugString()
+                    << " error_code : " << status.error_code()
+                    << " error_msg : " << status.error_message()
+                    << " source = " << source->toString();
+                return false;
+            }
+        },
+        ledgerSequence);
+    if (success)
+        return response;
+    else
+        return {};
+}
+
+std::unique_ptr<org::xrpl::rpc::v1::XRPLedgerAPIService::Stub>
+ETLLoadBalancer::getP2pForwardingStub()
+{
+    if (sources_.size() == 0)
+        return nullptr;
+    srand((unsigned)time(0));
+    auto sourceIdx = rand() % sources_.size();
+    auto numAttempts = 0;
+    while (numAttempts < sources_.size())
+    {
+        auto stub = sources_[sourceIdx]->getP2pForwardingStub();
+        if (!stub)
+        {
+            sourceIdx = (sourceIdx + 1) % sources_.size();
+            ++numAttempts;
+            continue;
+        }
+        return stub;
+    }
+    return nullptr;
+}
+
+Json::Value
+ETLLoadBalancer::forwardToP2p(RPC::JsonContext& context)
+{
+    Json::Value res;
+    if (sources_.size() == 0)
+        return res;
+    srand((unsigned)time(0));
+    auto sourceIdx = rand() % sources_.size();
+    auto numAttempts = 0;
+    while (numAttempts < sources_.size())
+    {
+        res = sources_[sourceIdx]->forwardToP2p(context);
+        if (!res.isMember("forwarded") || res["forwarded"] != true)
+        {
+            sourceIdx = (sourceIdx + 1) % sources_.size();
+            ++numAttempts;
+            continue;
+        }
+        return res;
+    }
+    RPC::Status err = {rpcFAILED_TO_FORWARD};
+    err.inject(res);
+    return res;
+}
+
+std::unique_ptr<org::xrpl::rpc::v1::XRPLedgerAPIService::Stub>
+ETLSource::getP2pForwardingStub()
+{
+    if (!connected_)
+        return nullptr;
+    try
+    {
+        return org::xrpl::rpc::v1::XRPLedgerAPIService::NewStub(
+            grpc::CreateChannel(
+                beast::IP::Endpoint(
+                    boost::asio::ip::make_address(ip_), std::stoi(grpcPort_))
+                    .to_string(),
+                grpc::InsecureChannelCredentials()));
+    }
+    catch (std::exception const& e)
```

Comment:
Nit: `e` is unreferenced. Consider removing it to silence the compiler warning.
e.
```
catch (std::exception const&)
```